src/coloring.rs

```diff
@@ -59,3 +67,209 @@ pub fn graph_greedy_color(py: Python, graph: &graph::PyGraph) -> PyResult<PyObje
     }
     Ok(out_dict.into())
 }
+
+struct MisraGriesAlgorithm<'a> {
+    graph: &'a StablePyGraph<Undirected>,
+    colors: DictMap<EdgeIndex, usize>,
+}
+
+impl<'a> MisraGriesAlgorithm<'a> {
+    pub fn new(graph: &'a StablePyGraph<Undirected>) -> Self {
+        let colors: DictMap<EdgeIndex, usize> = DictMap::new();
```

Comment:
I'd probably use `with_capacity` here instead of `new` since we know this will be `graph.edge_count()`. This will save us multiple memory allocations that would be made otherwise while it grows.